### PR TITLE
Solutions Chapter 4

### PR DESCRIPTION
**Pull request by L7R7**
### Solutions for Chapter 4

cc @vrom911 @chshersh Thank you very much for putting this together! I had a lot of fun solving all the exercises!

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +477,12 @@ Implement the Applicative instance for our 'Secret' data type from before.
...
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) _ (Trap e) = Trap e
+    (<*>) (Reward f) (Reward a) = Reward (f a)
```
Comment (vrom911) [inline src/Chapter4.hs]: This correct! 💯 
You can also notice that these two cases look very similar to the Functor instance of `Secret` 🗺️ 
You can actually reuse `fmap` 🙂

File: src/Chapter4.hs [hunk 2]
```diff
@@ -610,7 +627,19 @@ Implement the 'Monad' instance for our lists.
...
+  (>>=) as f = merge $ f <$> as
 
+merge :: List (List a) -> List a
+merge Empty = Empty
+merge (Cons as aas) = append as (merge aas)
+
+-- This could also be an instance of Append from Chapter 3
+append :: List a -> List a -> List a
+append Empty ys = ys
+append xs Empty = xs
```
Comment (vrom911) [inline src/Chapter4.hs:641]: Great catch!
Makes implementation much more efficient 💪🏼

File: src/Chapter4.hs [hunk 3]
```diff
@@ -629,7 +658,8 @@ Can you implement a monad version of AND, polymorphic over any monad?
 🕯 HINT: Use "(>>=)", "pure" and anonymous function
 -}
 andM :: (Monad m) => m Bool -> m Bool -> m Bool
-andM = error "andM: Not implemented!"
+-- andM mb mb' = mb >>= (\b -> (b &&) <$> mb')
+andM mb mb' = mb >>= \b -> if b then mb' else pure False -- coulnd't decide which one I like more. This is the lazy one if I understand correctly
```
Comment (vrom911) [inline src/Chapter4.hs:662]: Yes, this is the more correct one, as it is lazy, and we expected the lazy solution in here 🙂 
The ordinary `&&` if lazy on the second argument. Meaning that if it meets `False` as the first argument, there is no need to calculate another arg. In `andM` we would like to keep the same behaviour:slightly_smiling_face: 

In the first version of yours, you don't actually use the power of Monad and `>>=` in there could be easily replaced, saving the same further implementation. If you don't use the `>>=` operator, then the `Monad` constraint is too strong, and it is enough to specify only `Applicative`. However, with `Applicative` you are not able to implement a lazier version of `andM`.

File: src/Chapter4.hs [hunk 4]
```diff
@@ -672,7 +702,19 @@ Specifically,
...
+  fmap _ Leaf = Leaf
+  fmap f (Node a left right) = Node (f a) (fmap f left) (fmap f right)
+
+reverseTree :: Tree a -> Tree a
+reverseTree Leaf = Leaf
+reverseTree (Node a left right) = Node a (reverseTree right) (reverseTree left)
 
+toList :: Tree a -> List a -- traversing in-order
+toList Leaf = Empty
+toList (Node a left right) = append (toList left) (Cons a (toList right))
```
Comment (vrom911) [inline src/Chapter4.hs:717]: Great trees implementation 🌴

**Review by vrom911 — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! We are happy that you had fun 🥳 🍰  

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform) so we could improve in the future!

You can also reach us anytime after the course as well  :slightly_smiling_face: Don't be a stranger!

File: src/Chapter4.hs (hunk 1 above)
Comment (L7R7) [reply src/Chapter4.hs]: Oh, nice!

File: src/Chapter4.hs (hunk 3 above)
Comment (L7R7) [reply src/Chapter4.hs:662]: Yes, that's a good point. It's interesting to see that it is possible to implement it using `Applicative` exclusively, but only in a non-lazy version

**Comment by L7R7** [comment]
@vrom911 @chshersh Thanks again for the course and your feedback! I'm not really new to Haskell anymore, but I still learned quite a few things! :muscle: 

Let me also add that I really enjoy your blog posts, especially the one on mini patterns. Always a great resource to learn new things that can be applied in practice.
Next up is having a closer look at relude :wink: And I will give co-log another try as soon as the structured logging implementation found its way into the library, so I guess you will hear from me again in the future :D

**Pull request merged**